pkg/whip/sdk_media_sink.go

```diff
@@ -19,107 +19,67 @@ import (
-	"github.com/livekit/ingress/pkg/utils"
 	"github.com/livekit/protocol/livekit"
 	"github.com/livekit/protocol/logger"
 	"github.com/livekit/psrpc"
 )
 
 var (
 	ErrParamsUnavailable = psrpc.NewErrorf(psrpc.InvalidArgument, "codec parameters unavailable in sample")
+	ErrInvalidTracks     = psrpc.NewErrorf(psrpc.InvalidArgument, "lksdk returned an invalid number of tracks")
 )
 
 type SDKMediaSink struct {
-	logger     logger.Logger
-	params     *params.Params
-	writePLI   func()
-	track      *webrtc.TrackRemote
-	outputSync *utils.TrackOutputSynchronizer
-	sdkOutput  *lksdk_output.LKSDKOutput
+	logger   logger.Logger
+	params   *params.Params
+	writePLI func()
+	track    *webrtc.TrackRemote
 
-	readySamples     chan *sample
-	fuse             core.Fuse
-	trackInitialized bool
-}
+	sdkOutput  *lksdk_output.LKSDKOutput
+	localTrack *lksdk.LocalTrack
 
-type sample struct {
-	s  *media.Sample
-	ts time.Duration
+	fuse core.Fuse
 }
 
-func NewSDKMediaSink(l logger.Logger, p *params.Params, sdkOutput *lksdk_output.LKSDKOutput, track *webrtc.TrackRemote, outputSync *utils.TrackOutputSynchronizer, writePLI func()) *SDKMediaSink {
+func NewSDKMediaSink(l logger.Logger, p *params.Params, sdkOutput *lksdk_output.LKSDKOutput, track *webrtc.TrackRemote, writePLI func()) *SDKMediaSink {
 	s := &SDKMediaSink{
-		logger:       l,
-		params:       p,
-		writePLI:     writePLI,
-		track:        track,
-		outputSync:   outputSync,
-		sdkOutput:    sdkOutput,
-		readySamples: make(chan *sample, 15),
-		fuse:         core.NewFuse(),
+		logger:    l,
+		params:    p,
+		writePLI:  writePLI,
+		track:     track,
+		sdkOutput: sdkOutput,
+		fuse:      core.NewFuse(),
 	}
 
 	return s
 }
 
-func (sp *SDKMediaSink) PushSample(s *media.Sample, ts time.Duration) error {
+func (sp *SDKMediaSink) PushRTP(pkt *rtp.Packet) error {
 	if sp.fuse.IsBroken() {
 		return io.EOF
 	}
 
-	err := sp.ensureTrackInitialized(s)
-	if err != nil {
-		return err
-	}
-	if !sp.trackInitialized {
-		// Drop the sample
-		return nil
-	}
-
-	// Synchronize the outputs before the network jitter buffer to avoid old samples stuck
-	// in the channel from increasing the whole pipeline delay.
-	drop, err := sp.outputSync.WaitForMediaTime(ts)
-	if err != nil {
+	if err := sp.ensureTrackInitialized(pkt); err != nil {
 		return err
-	}
-	if drop {
-		sp.logger.Debugw("dropping sample", "timestamp", ts)
+	} else if sp.localTrack == nil {
+		// Drop the packet
 		return nil
 	}
 
-	select {
-	case <-sp.fuse.Watch():
-		return io.EOF
-	case sp.readySamples <- &sample{s, ts}:
-	default:
-		// drop the sample if the output queue is full. This is needed if we are reconnecting.
-	}
-
-	return nil
-}
-
-func (sp *SDKMediaSink) NextSample(ctx context.Context) (media.Sample, error) {
-	for {
-		select {
-		case <-sp.fuse.Watch():
-		case <-ctx.Done():
-			return media.Sample{}, io.EOF
-		case s := <-sp.readySamples:
-			return *s.s, nil
-		}
-	}
+	return sp.localTrack.WriteRTP(pkt, nil)
```

Comment:
Could you describe how A/V sync works in this case? My understanding of the workflow is that RTCP Sender Reports from the WHIP source are ignored in this non transcoded path. How will the go SDK generate Sender Reports in ingress? Will it use the time of day when `WriteRTP` is called? If so, I believe that any network jitter between the WHIP source and ingress, or any pipeline delay difference between the audio and video path in the source would affect A/V sync for the media relayed by ingress.